File: Python/Unique_Binary_Search_Trees.py
"""
Given n, how many structurally unique BST's (binary search trees) that store values 1...n?

For example,
Given n = 3, there are a total of 5 unique BST's.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
   
"""
class Solution:
    # @return an integer
    def numTrees(self, n):
        Num=[1,1]+[0]*n;
        for i in range(2,n+1):
            for j in range(i):
                Num[i]+=Num[j]*Num[i-j-1];
        return Num[n];

File: Java/src/Unique Binary Search Trees.java
public class Solution {
    public int numTrees(int n) {
        int[] Num = new int[n+1];
        Num[1] = Num[0] = 1;
        
        for (int i = 2; i <= n; i++ ){
            for (int j = 0; j<i; j++){
                Num[i]+= Num[j] * Num[i-1-j];
            }
        }
        return Num[n];
    }
}